### PR TITLE
Add support for structured curved mesh visualization in 2D

### PR DESCRIPTION
**Pull request by sloede**

### CODE REVIEW

**Comment by sloede** [comment]
@bennibolm @erik-f @NichtLucas 2D waved flag seems to work now. Starting `julia --project` in this branch, using
```julia
julia> using Trixi2Vtk

julia> Trixi.trixi2vts("../main-Trixi.jl/out/solution_000041.h5")
```
yields the following VTS visualization in ParaView: 

image

**Comment by sloede** [comment]
So, loading the mapping function from the string in the file is still somewhat brittle. With local ad-hoc modifications, I am able to also visualize the result of `examples/2d/elixir_euler_free_stream_curved.jl`:
image

**Comment by erik-f** [comment]
We now plot `CurvedMesh` data on Lagrange cells, which allows for perfect resolution with only `n_nodes` visualization nodes. However, the nonlinear subdivision level in ParaView needs to be turned up to get proper results.

We removed duplicate code and reused most of the code for `TreeMesh` data.

It is possible to use more visualization nodes than DG nodes, which is useful for curved cells. While ParaView shows the exact interpolation if the Lagrange nodes are equidistant, it shows different results otherwise. This means that the data on curved cells is not exactly the correct interpolation. Using more visualization nodes somehow mitigates this problem.

File: src/convert.jl [hunk 1]
```diff
@@ -97,30 +99,25 @@ function trixi2vtk(filename::AbstractString...;
 
     # Read mesh
     verbose && println("| Reading mesh file...")
-    @timeit "read mesh" (center_level_0, length_level_0,
-                         leaf_cells, coordinates, levels) = read_meshfile(meshfile)
+    # TODO 100_000 could be too small in some cases
+    @timeit "read mesh" mesh = Trixi.load_mesh_serial(meshfile; RealT=Float64, n_cells_max=100_000)
```
Comment (sloede) [inline src/convert.jl:103]:
```suggestion
    @timeit "read mesh" mesh = Trixi.load_mesh_serial(meshfile; RealT=Float64)
```
Can be omitted with Trixi v0.3.31

File: Project.toml [hunk 2]
```diff
@@ -17,5 +18,6 @@ HDF5 = "0.14, 0.15"
 ProgressMeter = "1.3"
 StaticArrays = "0.12, 1.0"
 TimerOutputs = "0.5"
+Trixi = "0.3.30"
```
Comment (sloede) [inline Project.toml:21]:
```suggestion
Trixi = "0.3.31"
```
to get rid of `n_cells_max`

File: src/vtktools.jl [hunk 3]
```diff
@@ -42,45 +44,190 @@ end
...
-                                                        vtk_celldata_points,
-                                                        vtk_celldata_cells)
+  end
+
+  # # Prepare VTK points and cells for celldata file
+  # @timeit "prepare VTK cells (cell data)" begin
+  #   vtk_celldata_points, vtk_celldata_cells = calc_vtk_points_cells(Val(ndims_), coordinates,
+  #                                                                   levels, center_level_0,
+  #                                                                   length_level_0, 1)
+  # end
```
Comment (sloede) [inline src/vtktools.jl:108]:
```suggestion
```
Obsolete?

File: src/vtktools.jl [hunk 4]
```diff
@@ -42,45 +44,190 @@ end
...
+  verbose && println("| Building VTK grid...")
+  if is_datafile
+    @timeit "build VTK grid (node data)" vtk_nodedata = vtk_grid(vtk_filename, vtk_points,
+                                                                 vtk_cells)
+  else
+    vtk_nodedata = nothing
+  end
+  # @timeit "build VTK grid (cell data)" vtk_celldata = vtk_grid(vtk_celldata_filename,
+  #                                                               vtk_celldata_points,
+  #                                                               vtk_celldata_cells)
```
Comment (sloede) [inline src/vtktools.jl:125]:
```suggestion
```
Obsolete?

File: src/vtktools.jl [hunk 5]
```diff
@@ -42,45 +44,190 @@ end
...
+      end
+      Trixi.calc_node_coordinates!(node_coordinates, element, nodes, four_corners)
+    end
+  end
+
+  return node_coordinates
+end
+
+
+function calc_node_coordinates!(node_coordinates::AbstractArray{<:Any, 4}, f, nodes, mesh)
```
Comment (sloede) [inline src/vtktools.jl:178]: AFAICT, this routine is only useful for the structured mesh and applies for 2D. Thus, I am wondering if one could mention this in a comment to this function (and for the next `calc_node_coordinates!`)

File: src/vtktools.jl [hunk 6]
```diff
@@ -263,3 +410,94 @@ function calc_vtk_points_cells(::Val{3}, coordinates::AbstractMatrix{Float64},
   return vtk_points, vtk_cells
 end
 
+
+# Convert coordinates and level information to a list of points and VTK cells (2D version)
```
Comment (sloede) [inline src/vtktools.jl:414]: This method and the next one are only used for the curved stuff, right? Could we please have a comment that mentions this?

**Review by sloede — COMMENTED** [review]
In general, I think this looks already pretty good!

@erik-f @bennibolm @NichtLucas I hope I can leave these changes to you?

File: src/convert.jl [hunk 7]
```diff
@@ -15,8 +15,10 @@ Convert Trixi-generated output files to VTK files (VTU or VTI).
 - `pvd`: Use this filename to store PVD file (instead of auto-detecting name). Note that
          only the name will be used (directory and file extension are ignored).
 - `output_directory`: Output directory where generated files are stored.
-- `nvisnodes`: Number of visualization nodes per element (default: twice the number of DG nodes).
+- `nvisnodes`: Number of visualization nodes per element.
+               (default: number of DG nodes for CurvedMesh, twice the number of DG nodes for TreeMesh).
```
Comment (andrewwinters5000) [inline src/convert.jl:19]:
```suggestion
               (default: number of DG nodes for CurvedMesh or UnstructuredQuadMesh,
                         twice the number of DG nodes for TreeMesh).
```

File: src/vtktools.jl [hunk 8]
```diff
@@ -42,45 +44,190 @@ end
...
+  end
+  @timeit "build VTK grid (cell data)" vtk_celldata = vtk_grid(vtk_celldata_filename,
+                                                      vtk_celldata_points,
+                                                      vtk_celldata_cells)
+
+  return vtk_nodedata, vtk_celldata
+end
+
+
+# Create and return VTK grids that are ready to be filled with data (CurvedMesh version)
```
Comment (andrewwinters5000) [inline src/vtktools.jl:89]:
```suggestion
# Create and return VTK grids that are ready to be filled with data (CurvedMesh/UnstructuredQuadMesh version)
```

**Review by andrewwinters5000 — CHANGES_REQUESTED** [review]

**Review by andrewwinters5000 — APPROVED** [review]

**Comment by bennibolm** [comment]
All suggestion should be implemented now. `n_max_cells=0` is still needed because we didn't add a default value there.

**Comment by sloede** [comment]
Can you please mark all resolved comments as such? And why do we still need n_cells_max=0?

**Comment by bennibolm** [comment]
I can't, it's not my PR. But all should be resolved now (maybe except the one with `n_cells_max`?).
In Trixi's [load_mesh_serial](https://github.com/trixi-framework/Trixi.jl/blob/ff549a5e67a7685f2ad3c97a0694c756160d79b4/src/mesh/mesh_io.jl#L150-160) we just added the `max(n_cells, n_cells_max)` but no default value for `n_cells_max`. We just added it for `load_mesh` but that's a restart file there.

**Pull request merged**